scripts/factory-common-install/bin/device-vnc-enter-as-frontend

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env bash
+set -euf -o pipefail
+common_factory_install_libexec_dir="$("$(dirname "$0")/pkg-nixos-factory-common-install-get-libexec-dir")"
+. "$common_factory_install_libexec_dir/app_current_device_ssh.sh"
+. "$common_factory_install_libexec_dir/app_current_device_store.sh"
+
+vnc_user=""
+vnc_hostname=""
+
+is_vnc_user_empty() {
+  if [ -z $vnc_user ]; then
+    echo "Error: The vnc user is not set"
+    return 2
+  fi
+}
+
+is_vnc_hostname_empty() {
+  if [ -z $vnc_hostname ]; then
+    echo "Error: The vnc hostname is not set"
+    return 2
+  fi
+}
+
+is_x11vnc_service_stopped() {
+  $(ssh -T ${vnc_user}@${vnc_hostname} systemctl -q --no-pager --user status nixos-secure-factory-x11vnc >/dev/null) && echo ok
+  err=${PIPESTATUS[0]}
+  if [ $err -ne 0 ]; then
+    if [ $err -eq 4 ]; then
+      echo "The service is not started yet, it's all good."
+      return 0
+    else
+      echo "Oops: Something went wrong. Error code: $err"
+      return $err
+    fi
+  fi
+}
+
+cleanup() {
+  echo "Cleanup required"
+  if ! is_x11vnc_service_stopped; then
+    $(ssh -T ${vnc_user}@${vnc_hostname} x11vnc -remote stop) && true
+    err=${PIPESTATUS[0]}
+    if [ $err -ne 0 ]; then
+      echo "error? $err"
+    fi
+  fi
+  exit
+}
+trap cleanup EXIT SIGINT SIGQUIT ERR
+
+enter_vnc_as_user() {
+  vnc_user="$1"
+  local local_port="$2"
+  local remote_port="$3"
+  local err
+  shift 3
+
+  read_or_prompt_for_current_device__hostname "vnc_hostname"
+  read_or_prompt_for_current_device__ssh_port "device_ssh_port"
+
+  if ! is_vnc_user_empty && ! is_vnc_hostname_empty; then
+    echo "Error: VNC user or hostname is not set"
+    exit 2
+  fi
+
+  # Validation that the service that we are about to launch does not exist yet.
+  echo " -> Making sure x11vnc is not aready started"
+  if ! is_x11vnc_service_stopped; then
+    cleanup
```

Comment:
Do we kick out an existing vnc session by doing that?

Reply:
The truth is we might be, but it shouldn't happens. If we wan't to make sure we are not kicking out an existing vnc session, I guess we should store the PID.

When I say it shouldn't happen is because of the following:
- x11VNC is launched with the once option so as soon one closes the communication the server is closed.
- *There shouldn't have two persons connecting at the same time*. Maybe we should had some kind of check?

Reply:
Couldn't we check the status of the systemd unit and issue return with an error if the vnc unit is alive and still running? If alive and still running best guest is that there is another vnc session open potentially by another user. We might also explicitly ask the user if he wants to kick out the existing user?

Reply:
In the case, the systemd unit is not alive and has exit with an error, we might assume that there was a failure to gc the unit in another run of the tool. We might issue a warning, gc the unit ourself and proceed.

Reply:
The current approach is to check if x11 is alive and if it is we issue the exit command. You are right, though we probably should ask the user first if he wants to kick the other user. Not sure if there is a way to retrieve information about how logged the last? What you think?

Reply:
As discussed, one way to implement this would be to wrap x11vnc and print details about the user who launched it (hostname, ip address, ssh fingerprint). Thus we should be able to see this when we call `systemctl status`.